python/simulation.py

```diff
@@ -2143,6 +2143,177 @@ def get_sfield_r(self):
+            elif direction == mp.Y and side == mp.Low: 
+                corner = np.array([self.geometry_center.x - self.cell_size.x/2,self.geometry_center.y + self.cell_size.y/2 - thickness])
+                width = self.cell_size.x
+                height = thickness
+            # Bottom
+            elif direction == mp.Y and side == mp.High: 
+                corner = np.array([self.geometry_center.x - self.cell_size.x/2,self.geometry_center.y - self.cell_size.y/2])
+                width = self.cell_size.x
+                height = thickness
+            else:
+                corner = []
+            return corner, width, height
+        
+        # All 4 side are the same
+        if boundary.direction == mp.ALL and boundary.side == mp.ALL:
+            for permutation in itertools.product([mp.X,mp.Y], [mp.Low, mp.High]):
+                corner, width, height = _get_boundary_vertices(boundary.thickness,*permutation)
+                corners.append(corner)
+                widths.append(width)
+                heights.append(height)
+        # 2 sides are the same
+        elif boundary.side == mp.ALL:
+             for side in [mp.Low, mp.High]:
+                corner, width, height = _get_boundary_vertices(boundary.thickness,boundary.direction,side)
+                corners.append(corner)
+                widths.append(width)
+                heights.append(height)
+        # only one side
+        else:
+            corners, widths, heights = _get_boundary_vertices(thickness, direction, side)
+        
+        return corners, widths, heights
+
+    def get_src_metadata(self,source):
+        size = source.size
+        center = source.center
+
+        # Point source
+        if size.x == size.y == size.z == 0:
+            return np.array([[center.x,center.y]])
+        
+        # Line source
+        else:
+            # Vertical line
+            if size.x == 0:
+                return np.array([
+                    [center.x,center.y-size.y/2],
+                    [center.x,center.y+size.y/2]
+                ])
+            # Horizontal line
+            elif size.y == 0:
+                return np.array([
+                    [center.x-size.x/2,center.y],
+                    [center.x+size.x/2,center.y]
+                ])
+            else:
+                return []
+            
+    def get_monitor_metadata(self,monitor,z_slice=0):
+        size = monitor.regions[0].size
+        center = monitor.regions[0].center
+
+        # Vertical line
+        if size.x == 0:
+            return np.array([
+                [center.x,center.y-size.y/2],
+                [center.x,center.y+size.y/2]
+            ])
+        # Horizontal line
+        elif size.y == 0:
+            return np.array([
+                [center.x-size.x/2,center.y],
+                [center.x+size.x/2,center.y]
+            ])
+        else:
+            return []
+
+    def visualize_domain(self,figsize=None,fields=None,labels=True):
```

Comment:
You could have keyword parameters `x=None, y=None, z=None`, so that you could pass e.g. `x=5` to specify the yz slice at `x=5`.  If all of them are `None`, defaulting to `z=0` is reasonable.

Reply:
We could have `3d=True` to plot a 3d visualization with Mayavi or whatever.